File: apd/src/supercall.rs
use std::{
    ffi::{CStr, CString},
    fs::File,
    io::{self, Read},
    process,
    sync::{Arc, Mutex},
};

use libc::{EINVAL, c_long, c_void, syscall, uid_t};
use log::{error, info, warn};

use crate::package::{read_ap_package_config, synchronize_package_uid};

const MAJOR: c_long = 0;
const MINOR: c_long = 11;
const PATCH: c_long = 1;

const KSTORAGE_EXCLUDE_LIST_GROUP: i32 = 1;

const __NR_SUPERCALL: c_long = 45;
const SUPERCALL_SU: c_long = 0x1010;
const SUPERCALL_KSTORAGE_WRITE: c_long = 0x1041;
const SUPERCALL_SU_GRANT_UID: c_long = 0x1100;
const SUPERCALL_SU_REVOKE_UID: c_long = 0x1101;
const SUPERCALL_SU_NUMS: c_long = 0x1102;
const SUPERCALL_SU_LIST: c_long = 0x1103;
const SUPERCALL_SU_RESET_PATH: c_long = 0x1111;
const SUPERCALL_SU_GET_SAFEMODE: c_long = 0x1112;

const SUPERCALL_SCONTEXT_LEN: usize = 0x60;

#[repr(C)]
struct SuProfile {
    uid: i32,
    to_uid: i32,
    scontext: [u8; SUPERCALL_SCONTEXT_LEN],
}

fn ver_and_cmd(cmd: c_long) -> c_long {
    let version_code: u32 = ((MAJOR << 16) + (MINOR << 8) + PATCH).try_into().unwrap();
    ((version_code as c_long) << 32) | (0x1158 << 16) | (cmd & 0xFFFF)
}

fn sc_su_revoke_uid(key: &CStr, uid: uid_t) -> c_long {
    if key.to_bytes().is_empty() {
        return (-EINVAL).into();
    }
    unsafe {
        syscall(
            __NR_SUPERCALL,
            key.as_ptr(),
            ver_and_cmd(SUPERCALL_SU_REVOKE_UID),
            uid,
        ) as c_long
    }
}

fn sc_su_grant_uid(key: &CStr, profile: &SuProfile) -> c_long {
    if key.to_bytes().is_empty() {
        return (-EINVAL).into();
    }
    unsafe {
        syscall(
            __NR_SUPERCALL,
            key.as_ptr(),
            ver_and_cmd(SUPERCALL_SU_GRANT_UID),
            profile,
        ) as c_long
    }
}

fn sc_kstorage_write(
    key: &CStr,
    gid: i32,
    did: i64,
    data: *mut c_void,
    offset: i32,
    dlen: i32,
) -> c_long {
    if key.to_bytes().is_empty() {
        return (-EINVAL).into();
    }
    unsafe {
        syscall(
            __NR_SUPERCALL,
            key.as_ptr(),
            ver_and_cmd(SUPERCALL_KSTORAGE_WRITE),
            gid as c_long,
            did as c_long,
            data,
            (((offset as i64) << 32) | (dlen as i64)) as c_long,
        ) as c_long
    }
}

fn sc_set_ap_mod_exclude(key: &CStr, uid: i64, exclude: i32) -> c_long {
    sc_kstorage_write(
        key,
        KSTORAGE_EXCLUDE_LIST_GROUP,
        uid,
        &exclude as *const i32 as *mut c_void,
        0,
        size_of::<i32>() as i32,
    )
}

pub fn sc_su_get_safemode(key: &CStr) -> c_long {
    if key.to_bytes().is_empty() {
        warn!("[sc_su_get_safemode] null superkey, tell apd we are not in safemode!");
        return 0;
    }

    let key_ptr = key.as_ptr();
    if key_ptr.is_null() {
        warn!("[sc_su_get_safemode] superkey pointer is null!");
        return 0;
    }

    unsafe {
        syscall(
            __NR_SUPERCALL,
            key_ptr,
            ver_and_cmd(SUPERCALL_SU_GET_SAFEMODE),
        ) as c_long
    }
}

fn sc_su(key: &CStr, profile: &SuProfile) -> c_long {
    if key.to_bytes().is_empty() {
        return (-EINVAL).into();
    }
    unsafe {
        syscall(
            __NR_SUPERCALL,
            key.as_ptr(),
            ver_and_cmd(SUPERCALL_SU),
            profile,
        ) as c_long
    }
}

fn sc_su_reset_path(key: &CStr, path: &CStr) -> c_long {
    if key.to_bytes().is_empty() || path.to_bytes().is_empty() {
        return (-EINVAL).into();
    }
    unsafe {
        syscall(
            __NR_SUPERCALL,
            key.as_ptr(),
            ver_and_cmd(SUPERCALL_SU_RESET_PATH),
            path.as_ptr(),
        ) as c_long
    }
}



fn sc_su_uid_nums(key: &CStr) -> c_long {
    if key.to_bytes().is_empty() {
        return (-EINVAL).into();
    }
    unsafe { syscall(__NR_SUPERCALL, key.as_ptr(), ver_and_cmd(SUPERCALL_SU_NUMS)) as c_long }
}

fn sc_su_allow_uids(key: &CStr, buf: &mut [uid_t]) -> c_long {
    if key.to_bytes().is_empty() {
        return (-EINVAL).into();
    }
    if buf.is_empty() {
        return (-EINVAL).into();
    }
    unsafe {
        syscall(
            __NR_SUPERCALL,
            key.as_ptr(),
            ver_and_cmd(SUPERCALL_SU_LIST),
            buf.as_mut_ptr(),
            buf.len() as i32,
        ) as c_long
    }
}

fn read_file_to_string(path: &str) -> io::Result<String> {
    let mut file = File::open(path)?;
    let mut content = String::new();
    file.read_to_string(&mut content)?;
    Ok(content)
}

fn convert_string_to_u8_array(s: &str) -> [u8; SUPERCALL_SCONTEXT_LEN] {
    let mut u8_array = [0u8; SUPERCALL_SCONTEXT_LEN];
    let bytes = s.as_bytes();
    let len = usize::min(SUPERCALL_SCONTEXT_LEN, bytes.len());
    u8_array[..len].copy_from_slice(&bytes[..len]);
    u8_array
}

fn convert_superkey(s: &Option<String>) -> Option<CString> {
    s.as_ref().and_then(|s| CString::new(s.clone()).ok())
}

pub fn refresh_ap_package_list(skey: &CStr, mutex: &Arc<Mutex<()>>) {
    let _lock = mutex.lock().unwrap();

    let num = sc_su_uid_nums(skey);
    if num < 0 {
        error!("[refresh_su_list] Error getting number of UIDs: {}", num);
        return;
    }
    let num = num as usize;
    let mut uids = vec![0 as uid_t; num];
    let n = sc_su_allow_uids(skey, &mut uids);
    if n < 0 {
        error!("[refresh_su_list] Error getting su list");
        return;
    }
    for uid in &uids {
        if *uid == 0 || *uid == 2000 {
            warn!(
                "[refresh_ap_package_list] Skip revoking critical uid: {}",
                uid
            );
            continue;
        }
        info!(
            "[refresh_ap_package_list] Revoking {} root permission...",
            uid
        );
        let rc = sc_su_revoke_uid(skey, *uid);
        if rc != 0 {
            error!("[refresh_ap_package_list] Error revoking UID: {}", rc);
        }
    }

    if let Err(e) = synchronize_package_uid() {
        error!("Failed to synchronize package UIDs: {}", e);
    }

    let package_configs = read_ap_package_config();
    for config in package_configs {
        if config.allow == 1 && config.exclude == 0 {
            let profile = SuProfile {
                uid: config.uid,
                to_uid: config.to_uid,
                scontext: convert_string_to_u8_array(&config.sctx),
            };
            let result = sc_su_grant_uid(skey, &profile);
            info!(
                "[refresh_ap_package_list] Loading {}: result = {}",
                config.pkg, result
            );
        }
        if config.allow == 0 && config.exclude == 1 {
            let result = sc_set_ap_mod_exclude(skey, config.uid as i64, 1);
            info!(
                "[refresh_ap_package_list] Loading exclude {}: result = {}",
                config.pkg, result
            );
        }
    }
}

pub fn privilege_apd_profile(superkey: &Option<String>) {
    let key = convert_superkey(superkey);

    let all_allow_ctx = "u:r:magisk:s0";
    let profile = SuProfile {
        uid: process::id().try_into().expect("PID conversion failed"),
        to_uid: 0,
        scontext: convert_string_to_u8_array(all_allow_ctx),
    };
    if let Some(ref key) = key {
        let result = sc_su(key, &profile);
        info!("[privilege_apd_profile] result = {}", result);
    }
}


pub fn init_load_su_path(superkey: &Option<String>) {
    let su_path_file = "/data/adb/ap/su_path";

    match read_file_to_string(su_path_file) {
        Ok(su_path) => {
            let superkey_cstr = convert_superkey(superkey);

            match superkey_cstr {
                Some(superkey_cstr) => match CString::new(su_path.trim()) {
                    Ok(su_path_cstr) => {
                        let result = sc_su_reset_path(&superkey_cstr, &su_path_cstr);
                        if result == 0 {
                            info!("suPath load successfully");
                        } else {
                            warn!("Failed to load su path, error code: {}", result);
                        }
                    }
                    Err(e) => {
                        warn!("Failed to convert su_path: {}", e);
                    }
                },
                _ => {
                    warn!("Superkey is None, skipping...");
                }
            }
        }
        Err(e) => {
            warn!("Failed to read su_path file: {}", e);
        }
    }
}

File: apd_go/supercall.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
	"syscall"
	"unsafe"
)

type SuProfile struct {
	Uid      int32
	ToUid    int32
	Scontext [SUPERCALL_SCONTEXT_LEN]byte
}

func verAndCmd(key string, cmd int64) int64 {
	versionCode := (MAJOR << 16) | (MINOR << 8) | PATCH
	return (int64(versionCode) << 32) | (0x1158 << 16) | (cmd & 0xFFFF)
}
func indexNullByte(b []byte) int {
	for i, v := range b {
		if v == 0 {
			return i
		}
	}
	return -1
}
func convertStringToSctx(s string) [256]byte {
	var b [256]byte
	copy(b[:], s)
	return b
}
func test(key string) {
	buf := make([]byte, 256)     // 路径缓冲区
	ret := scSuGetPath(key, buf) // 调用超级调用

	if ret < 0 {
		fmt.Printf("scSuGetPath failed: %d\n", ret)
		return
	}
	path := string(buf)
	if i := indexNullByte(buf); i >= 0 {
		path = string(buf[:i])
	}
	fmt.Printf("Path: %s\n", path)
}

// test function
func test1(key string) {
	var profile SuProfile
	profile.Uid = 1000                         // 当前用户 UID
	profile.ToUid = 2000                       // 目标用户 UID
	copy(profile.Scontext[:], "untrusted_app") // 填充安全上下文

	result := supercall(key, unsafe.Pointer(&profile))
	fmt.Printf("Result: %d\n", result)
}

func supercall(key string, profilePtr unsafe.Pointer) int64 {
	if len(key) == 0 {
		return -int64(syscall.EINVAL)
	}

	cKey := []byte(key + "\x00")
	keyPtr := unsafe.Pointer(&cKey[0])

	ret, _, errno := syscall.RawSyscall(
		uintptr(__NR_SUPERCALL),
		uintptr(keyPtr),
		uintptr(SUPERCALL_SU),
		uintptr(profilePtr),
	)

	if errno != 0 {
		return int64(-errno)
	}

	return int64(ret)
}

func scSuGrantUid(key string, profile *SuProfile) int64 {
	if len(key) == 0 {
		return -int64(syscall.EINVAL)
	}

	cKey := append([]byte(key), 0)
	keyPtr := unsafe.Pointer(&cKey[0])

	ret, _, errno := syscall.RawSyscall(
		uintptr(__NR_SUPERCALL),
		uintptr(keyPtr),
		uintptr(verAndCmd(key, SUPERCALL_SU_GRANT_UID)),
		uintptr(unsafe.Pointer(profile)),
	)

	if errno != 0 {
		return -int64(errno)
	}
	return int64(ret)
}

func scSuGetPath(key string, outPath []byte) int64 {
	pathLen := len(outPath)

	if len(key) == 0 {
		return -int64(syscall.EINVAL)
	}
	if outPath == nil || pathLen <= 0 {
		return -int64(syscall.EINVAL)
	}

	cKey := append([]byte(key), 0)
	keyPtr := unsafe.Pointer(&cKey[0])

	outPtr := unsafe.Pointer(&outPath[0])

	ret, _, errno := syscall.RawSyscall6(
		uintptr(__NR_SUPERCALL),
		uintptr(keyPtr),
		uintptr(verAndCmd(key, SUPERCALL_SU_GET_PATH)),
		uintptr(outPtr),
		uintptr(pathLen),
		0,
		0,
	)

	if errno != 0 {
		return -int64(errno)
	}
	return int64(ret)
}

func scKstorageWrite(key string, gid int32, did int64, data unsafe.Pointer, offset int32, dlen int32) int64 {
	if len(key) == 0 {
		return -int64(syscall.EINVAL)
	}

	cKey := append([]byte(key), 0)
	keyPtr := unsafe.Pointer(&cKey[0])

	offsetDlen := int64(offset)<<32 | int64(dlen)

	ret, _, errno := syscall.RawSyscall6(
		uintptr(__NR_SUPERCALL),
		uintptr(keyPtr),
		uintptr(verAndCmd(key, SUPERCALL_KSTORAGE_WRITE)),
		uintptr(gid),
		uintptr(did),
		uintptr(data),
		uintptr(offsetDlen),
	)

	if errno != 0 {
		return -int64(errno)
	}
	return int64(ret)
}

func InitLoadPackageUidConfig(superkey string) {
	packageConfigs := ReadApPackageConfig()

	for _, config := range packageConfigs {
		if config.Allow == 1 && config.Exclude == 0 {
			if len(superkey) > 0 {
				profile := SuProfile{
					Uid:      int32(config.Uid),
					ToUid:    int32(config.ToUid),
					Scontext: convertStringToSctx(config.Sctx),
				}
				result := scSuGrantUid(superkey, &profile)
				Info("Processed %s: result = %d\n", config.Pkg, result)
			} else {
				Error("Superkey is None, skipping config: %s\n", config.Pkg)
			}
		}

		if config.Allow == 0 && config.Exclude == 1 {
			if len(superkey) > 0 {
				result := scSetApModExclude(superkey, int64(config.Uid), 1)
				Info("Processed exclude %s: result = %d\n", config.Pkg, result)
			} else {
				Error("Superkey is None, skipping config: %s\n", config.Pkg)
			}
		}
	}
}

func scSetApModExclude(key string, uid int64, exclude int32) int64 {
	return scKstorageWrite(
		key,
		KSTORAGE_EXCLUDE_LIST_GROUP,   // gid
		uid,                           // did
		unsafe.Pointer(&exclude),      // data
		0,                             // offset
		int32(unsafe.Sizeof(exclude)), // dlen
	)
}

func scSuResetPath(key string, suPath string) int64 {
	if len(key) == 0 || len(suPath) == 0 {
		return -1
	}

	cKey := append([]byte(key), 0)
	keyPtr := unsafe.Pointer(&cKey[0])

	cPath := append([]byte(suPath), 0)
	pathPtr := unsafe.Pointer(&cPath[0])

	ret, _, errno := syscall.RawSyscall6(
		uintptr(__NR_SUPERCALL),
		uintptr(keyPtr),
		uintptr(verAndCmd(key, SUPERCALL_SU_RESET_PATH)),
		uintptr(pathPtr),
		0, 0, 0,
	)
	if errno != 0 {
		return -int64(errno)
	}
	return int64(ret)
}
func InitLoadSuPath(superkey string) {
	suPathFile := "/data/adb/ap/su_path"

	file, err := os.Open(suPathFile)
	if err != nil {
		Error("Failed to read su_path file: %v\n", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var suPath string
	if scanner.Scan() {
		suPath = strings.TrimSpace(scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		Error("Failed to read su_path file: %v\n", err)
		return
	}

	if len(superkey) == 0 {
		Error("Superkey is None, skipping...")
		return
	}

	if len(suPath) == 0 {
		Error("su_path is empty, skipping...")
		return
	}

	result := scSuResetPath(superkey, suPath)
	if result == 0 {
		Info("suPath load successfully")
	} else {
		Info("Failed to load su path, error code: %d\n", result)
	}
}
func scSu(key string, profile *SuProfile) int64 {
	if len(key) == 0 {
		return -int64(syscall.EINVAL)
	}

	cKey := append([]byte(key), 0)
	keyPtr := unsafe.Pointer(&cKey[0])

	ret, _, errno := syscall.RawSyscall(
		uintptr(__NR_SUPERCALL),
		uintptr(keyPtr),
		uintptr(verAndCmd(key, SUPERCALL_SU)),
		uintptr(unsafe.Pointer(profile)),
	)

	if errno != 0 {
		return -int64(errno)
	}
	return int64(ret)
}

func PrivilegeApdProfile(key string) {
	if key == "" {
		log.Println("[PrivilegeApdProfile] empty key, skipping privilege")
		return
	}
	var sctx [256]byte
	copy(sctx[:], []byte("u:r:magisk:s0"))
	profile := SuProfile{
		Uid:      int32(os.Getpid()),
		ToUid:    0,
		Scontext: sctx,
	}

	result := scSu(key, &profile)
	Info("[PrivilegeApdProfile] result = %d\n", result)
}
func scSuGetSafemode(key string) int64 {
	if len(key) == 0 {
		return 0
	}

	cKey := append([]byte(key), 0)
	keyPtr := unsafe.Pointer(&cKey[0])

	ret, _, errno := syscall.RawSyscall(
		uintptr(__NR_SUPERCALL),
		uintptr(keyPtr),
		uintptr(verAndCmd(key, SUPERCALL_SU_GET_SAFEMODE)),
		0,
	)

	if errno != 0 {
		return -int64(errno)
	}
	return int64(ret)
}
